File: dump.py
#!/usr/bin/env python3

import os
import shutil
import tempfile
from pathlib import Path
from urllib.request import Request, urlopen
from zipfile import ZipFile

URL = "https://file.garden/ZzSF-2oH_UI6UVla/streamable.com.zip"
DEST = Path("dump").resolve()

DEST.mkdir(parents=True, exist_ok=True)

headers = {
    "User-Agent": "Mozilla/5.0",
    "Accept": "*/*",
}

def long_path(path: Path) -> str:
    resolved = str(path.resolve())

    if os.name == "nt":
        if resolved.startswith("\\\\"):
            return "\\\\?\\UNC\\" + resolved[2:]
        return "\\\\?\\" + resolved

    return resolved


with tempfile.TemporaryDirectory() as tmp:
    zip_path = Path(tmp) / "streamable.com.zip"

    print(f"Downloading {URL}...")

    request = Request(URL, headers=headers)

    with urlopen(request) as response, open(zip_path, "wb") as f:
        shutil.copyfileobj(response, f)

    print(f"Extracting to {DEST}...")

    with ZipFile(zip_path) as archive:
        for member in archive.infolist():
            target = DEST / member.filename

            try:
                target.resolve().relative_to(DEST)
            except ValueError:
                raise RuntimeError(f"Unsafe path in ZIP: {member.filename}")

            target_str = long_path(target)

            if member.is_dir():
                os.makedirs(target_str, exist_ok=True)
                continue

            parent_str = long_path(target.parent)
            os.makedirs(parent_str, exist_ok=True)

            with archive.open(member) as src, open(target_str, "wb") as dst:
                shutil.copyfileobj(src, dst)

print("Done.")

File: scripts/dump.js
#!/usr/bin/env bun

import { mkdir, rm, stat } from 'node:fs/promises';
import { dirname, isAbsolute, relative, resolve, sep } from 'node:path';
import { unzipSync } from 'fflate';

const DOWNLOAD_URL = 'https://file.garden/ZzSF-2oH_UI6UVla/streamable.com.zip';
const DUMP_DIRECTORY = resolve(import.meta.dir, '..', 'dump');
const DOWNLOAD_URL_FILE = resolve(DUMP_DIRECTORY, 'download-url.txt');

export function longPath(path, platform = process.platform) {
    if (platform !== 'win32') {
        return path;
    }
    if (path.startsWith('\\\\')) {
        return `\\\\?\\UNC\\${path.slice(2)}`;
    }
    return `\\\\?\\${path}`;
}

async function isDirectory(path) {
    try {
        return (await stat(longPath(path))).isDirectory();
    } catch {
        return false;
    }
}

async function hasCurrentDump() {
    if (!(await isDirectory(DUMP_DIRECTORY))) {
        return false;
    }

    const downloadUrlFile = Bun.file(longPath(DOWNLOAD_URL_FILE));
    return (
        (await downloadUrlFile.exists()) &&
        (await downloadUrlFile.text()).trim() === DOWNLOAD_URL
    );
}

function dumpPath(entryName) {
    const path = resolve(DUMP_DIRECTORY, entryName);
    const pathFromDump = relative(DUMP_DIRECTORY, path);

    if (
        pathFromDump === '..' ||
        pathFromDump.startsWith(`..${sep}`) ||
        isAbsolute(pathFromDump)
    ) {
        throw new Error(`Unsafe path in ZIP: ${entryName}`);
    }

    return path;
}

async function main() {
    if (await hasCurrentDump()) {
        console.log(`Dump is current: ${DOWNLOAD_URL}`);
        return;
    }

    await rm(longPath(DUMP_DIRECTORY), { recursive: true, force: true });

    try {
        console.log(`Downloading ${DOWNLOAD_URL}...`);
        const response = await fetch(DOWNLOAD_URL, {
            headers: {
                Accept: '*/*',
                'User-Agent': 'Mozilla/5.0'
            }
        });

        if (!response.ok) {
            throw new Error(
                `Download failed with HTTP ${response.status} ${response.statusText}`
            );
        }

        const archive = new Uint8Array(await response.arrayBuffer());
        const entries = unzipSync(archive);

        console.log(`Extracting to ${DUMP_DIRECTORY}...`);
        await mkdir(longPath(DUMP_DIRECTORY), { recursive: true });

        for (const [entryName, contents] of Object.entries(entries)) {
            if (entryName.endsWith('/')) {
                continue;
            }

            const path = dumpPath(entryName);
            await mkdir(longPath(dirname(path)), { recursive: true });
            await Bun.write(longPath(path), contents);
        }

        await Bun.write(longPath(DOWNLOAD_URL_FILE), `${DOWNLOAD_URL}\n`);
        console.log('Done.');
    } catch (error) {
        await rm(longPath(DUMP_DIRECTORY), { recursive: true, force: true });
        throw error;
    }
}

if (import.meta.main) {
    await main();
}
